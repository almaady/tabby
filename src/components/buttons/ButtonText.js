import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

/**
 * @class ButtonText - Is a button that only contains text
 */
class ButtonText extends React.PureComponent {
  /**
   * @property {string} text -The text of the button
   * @property {func} onClick - The function triggered by the button
   * @property {string} className - Just in case you need another class
   * @property {string} color -  The color of the button
   * @property {string} type - Button type
   * @property {bool} disabled - Button could be disabled
   * @property {string} borderColor - Button could have a border color
   * @see See './../../../styles/variables/_colors.scss' for the list of color names
   */
  static propTypes = {
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    color: PropTypes.string.isRequired,
    className: PropTypes.string,
    type: PropTypes.string,
    disabled: PropTypes.bool,
    borderColor: PropTypes.string
  };


  render() {
    const { color, onClick, text, className, type, disabled, borderColor } = this.props;
    return (
      <button
        className={classNames("btn", "btn--text", `bg--${color}`, `border--${borderColor}`,className)}
        onClick={onClick}
        type={type}
        disabled={disabled}
      >
        <p>{text}</p>
      </button>
    );
  }
}
ButtonText.defaultProps = {
  className: "",
  type: "button",
  disabled: false,
  borderColor: ""
};

export default ButtonText;
