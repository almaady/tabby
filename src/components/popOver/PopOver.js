// PopOver
import React from "react";
import PropTypes from "prop-types";
import PopOverCard from "./PopOverCard";
import { popXPosition, popYPosition } from "../../utils/popsPosition";

/**
 * @class PopOver - It contains the pop over card and its activator
 */
class PopOver extends React.PureComponent {
  /**
   * @property {node} children - The content of the pop over
   * @property {string} className - Just in case you need another class
   * @property {func} onClose - The function to close the pop over
   * @property {node} activator - The button/link that triggers the pop over
   * @property {bool} show - Determines if the popOver is visible
   * @property {string} title - The title of the pop over
   * @property {bool} back - Determines if the pop over has a back button
   */

  static propTypes = {
    activator: PropTypes.node.isRequired,
    onClose: PropTypes.func.isRequired,
    children: PropTypes.node.isRequired,
    show: PropTypes.bool.isRequired,
    title: PropTypes.string,
    back: PropTypes.bool,
    className: PropTypes.string
  };

  static defaultProps = {
    title: "",
    back: false,
    className: ""
  };

  constructor(props) {
    super(props);
    this.activatorRef = React.createRef();
    this.popRef = React.createRef();
  }

  componentDidMount() {
    document.addEventListener("mousedown", this.handleClickOutside);
    window.addEventListener("resize", this.resize);
    document.addEventListener("keydown", this.handleEscPressed);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.handleClickOutside);
    window.removeEventListener("resize", this.resize);
    document.removeEventListener("keydown", this.handleEscPressed);
  }

  isSafari(){
    let ua = navigator.userAgent.toLowerCase(); 
    if (ua.indexOf('safari') != -1) { 
      if (ua.indexOf('chrome') === -1) {
        return true;
      }
    }
    return false;
  }

  getStyle() {
    const activator = this.activatorRef.current;
    if (!activator || this.isSafari()) {
      return {};
    }
    const pop = this.popRef.current;
    const leftPosition = popXPosition(activator, pop);
    const topPosition = popYPosition(activator, pop);
    return {
      left: `${leftPosition}px`,
      top: `${topPosition}px`
    };
  }

  resize = () => this.forceUpdate();

  handleEscPressed = event => {
    const { onClose } = this.props;
    // keyCode 27 = ESC key
    if (event.keyCode === 27) {
      onClose();
    }
  };

  handleClickOutside = event => {
    if (
      this.popRef.current &&
      !this.popRef.current.contains(event.target) &&
      this.props.show
    ) {
      this.props.onClose();
    }
  };

  render() {
    const {
      onClose,
      children,
      className,
      activator,
      title,
      back,
      onBack,
      show
    } = this.props;
    const { left, top } = this.getStyle();
    return (
      // eslint-disable-next-line react/jsx-filename-extension
      <div className="pop-over--wrapper" ref={this.popRef}>
        <div ref={this.activatorRef} id="activator">
          {activator}
        </div>
        <PopOverCard
          onClose={onClose}
          className={className}
          show={show}
          left={left}
          top={top}
          title={title}
          back={back}
          onBack={onBack}
        >
          {children}
        </PopOverCard>
      </div>
    );
  }
}

export default PopOver;
