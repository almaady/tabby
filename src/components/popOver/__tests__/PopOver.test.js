import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";

import PopOver from "../PopOver";
import ButtonText from "../../buttons/ButtonText";

describe("PopOver", () => {
  const close = sinon.stub();
  let props;
  beforeEach(() => {
    props = {
      show: false,
      children: <div/>,
      onClose: close,
      activator: <ButtonText color="blue" onClick={() => {}} text="Activator" />
    }
  });

  it("should render correctly with default state", () => {
    const component = shallow(<PopOver {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should call onClose when you press ESC key ", () => {
    shallow(<PopOver {...props} />);
    // eslint-disable-next-line no-undef
    const event = new KeyboardEvent("keydown", { keyCode: 27 });
    document.dispatchEvent(event);
    expect(close.called).toBeTruthy();
  });
});