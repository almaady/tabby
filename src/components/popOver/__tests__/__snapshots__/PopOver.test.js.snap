// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopOver should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PopOver
    activator={
      <ButtonText
        borderColor=""
        className=""
        color="blue"
        disabled={false}
        onClick={[Function]}
        text="Activator"
        type="button"
      />
    }
    back={false}
    className=""
    onClose={[Function]}
    show={false}
    title=""
  >
    <div />
  </PopOver>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          id="activator"
        >
          <ButtonText
            borderColor=""
            className=""
            color="blue"
            disabled={false}
            onClick={[Function]}
            text="Activator"
            type="button"
          />
        </div>,
        <PopOverCard
          back={false}
          className=""
          left=""
          onBack={null}
          onClose={[Function]}
          show={false}
          title=""
          top=""
        >
          <div />
        </PopOverCard>,
      ],
      "className": "pop-over--wrapper",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ButtonText
            borderColor=""
            className=""
            color="blue"
            disabled={false}
            onClick={[Function]}
            text="Activator"
            type="button"
          />,
          "id": "activator",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "borderColor": "",
            "className": "",
            "color": "blue",
            "disabled": false,
            "onClick": [Function],
            "text": "Activator",
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "back": false,
          "children": <div />,
          "className": "",
          "left": "",
          "onBack": null,
          "onClose": [Function],
          "show": false,
          "title": "",
          "top": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            id="activator"
          >
            <ButtonText
              borderColor=""
              className=""
              color="blue"
              disabled={false}
              onClick={[Function]}
              text="Activator"
              type="button"
            />
          </div>,
          <PopOverCard
            back={false}
            className=""
            left=""
            onBack={null}
            onClose={[Function]}
            show={false}
            title=""
            top=""
          >
            <div />
          </PopOverCard>,
        ],
        "className": "pop-over--wrapper",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ButtonText
              borderColor=""
              className=""
              color="blue"
              disabled={false}
              onClick={[Function]}
              text="Activator"
              type="button"
            />,
            "id": "activator",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "borderColor": "",
              "className": "",
              "color": "blue",
              "disabled": false,
              "onClick": [Function],
              "text": "Activator",
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "back": false,
            "children": <div />,
            "className": "",
            "left": "",
            "onBack": null,
            "onClose": [Function],
            "show": false,
            "title": "",
            "top": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
